### PR TITLE
Make s3 backup optional.

### PR DESCRIPTION
**Pull request by ndokos**
If the S3 service is inaccessible, the backup does not succeed and
keeps retrying. That may be reasonable if the S3 service is up and
running, but there may be situations where we want to bypass it.

Commenting out the [pbench-server-backup] section (or just the
endpoint_url option in that section) in the server config file now
allows the rest of the backup to succeed (including moving the
TO-BACKUP link to the final BACKED-UP state). That allows us to do
local backups, irrespective of the S3 backup service.

test-6.13 implements the commenting out of the endpoint_url option;
test-6.14 implements the commenting out of the [pbench-server-backup]
section (the tar ball is a symlink to the 6.13 tarball).

### CODE REVIEW

File: server/lib/s3backup/__init__.py [hunk 1]
```diff
@@ -64,6 +63,11 @@ def __init__(self, config, logger):
         else:
             self.connector = S3Connector(config)
 
+        try:
+            self.bucket_name = self.connector.bucket_name
+        except Exception:
+            pass
```
Comment (portante) [inline server/lib/s3backup/__init__.py]: Can we add a comment explaining why we can `pass` here and ignore the exception?

File: server/lib/s3backup/__init__.py [hunk 2]
```diff
@@ -206,19 +210,25 @@ def getsize(self, tb):
...
-            'pbench-server-backup', 'access_key_id')
-        self.secret_access_key = config.get(
-            'pbench-server-backup', 'secret_access_key')
-        self.s3client = boto3.client('s3',
-                                     aws_access_key_id='{}'.format(
-                                         self.access_key_id),
-                                     aws_secret_access_key='{}'.format(
-                                         self.secret_access_key),
-                                     endpoint_url='{}'.format(self.endpoint_url))
-        self.transfer_config = None
```
Comment (portante) [inline server/lib/s3backup/__init__.py]: We no longer need the `transfer_config` attribute?

File: server/lib/s3backup/__init__.py [hunk 3]
```diff
@@ -334,6 +350,7 @@ def upload_fileobj(self, Body=None, Bucket=None, Key=None, Config=None, ExtraArg
                                        ExtraArgs['Metadata']['MD5']))
 
     def head_bucket(self, Bucket=None):
+
```
Comment (portante) [inline server/lib/s3backup/__init__.py]: Is this addition intentional?

File: server/bin/pbench-backup-tarballs.py [hunk 4]
```diff
@@ -105,6 +105,10 @@ def sanity_check(lb_obj, s3_obj, config, logger):
     # make sure the S3 bucket exists
     try:
         s3_obj.head_bucket(s3_obj.bucket_name)
+    except AttributeError:
+        logger.error(
+            "Bucket: not defined in config file")
```
Comment (portante) [inline server/bin/pbench-backup-tarballs.py]: Can you make this message a `warning` and add to it saying that S3 backups are not enabled as a result?

File: server/bin/state/test-6.13.config/pbench-server.cfg [hunk 5]
```diff
@@ -0,0 +1,54 @@
+install-dir = %(unittest-dir)s/opt/pbench-server
```
Comment (portante) [inline server/bin/state/test-6.13.config/pbench-server.cfg]: What is the difference between the `test-6.13.config/pbench-server.cfg` file and the `test-6.14` one?  And further, can this be symlinks to existing files that are the same?

File: server/lib/s3backup/__init__.py (hunk 2 above)
Comment (ndokos) [reply server/lib/s3backup/__init__.py]: We do not.

File: server/lib/s3backup/__init__.py (hunk 1 above)
Comment (ndokos) [reply server/lib/s3backup/__init__.py]: Yes, I'll do that.

File: server/lib/s3backup/__init__.py (hunk 3 above)
Comment (ndokos) [reply server/lib/s3backup/__init__.py]: I don't think so. I'll revert it.

File: server/bin/pbench-backup-tarballs.py (hunk 4 above)
Comment (ndokos) [reply server/bin/pbench-backup-tarballs.py]: Sure.

File: server/bin/pbench-backup-tarballs.py [hunk 6]
```diff
@@ -105,9 +105,13 @@ def sanity_check(lb_obj, s3_obj, config, logger):
...
         s3_obj.head_bucket(s3_obj.bucket_name)
+    except AttributeError:
+        logger.warning(
+            "Bucket: not defined in config file - S3 backup is disabled.")
+        s3_obj = None
     except Exception:
-        logger.error(
-            "Bucket: {} does not exist or you have no access".format(s3_obj.bucket_name))
+        logger.warning(
+            "Bucket: {} is not accesssible - S3 backup is disabled".format(s3_obj.bucket_name))
```
[the remark this answers is not in the document]
Comment (ndokos) [reply server/bin/pbench-backup-tarballs.py]: What, you don't speak parseltongue?

File: server/lib/s3backup/__init__.py [hunk 7]
```diff
@@ -64,6 +63,25 @@ def __init__(self, config, logger):
...
+        # not defined at all (which is possible when we deliberately
+        # turn off the S3 backup by deleting the relevant section in
+        # the config file), we would get an AttributeError exception
+        # here - but we don't want to abort the application: it might
+        # still want to continue with local backups, so we ignore the
+        # exception: the application will check it.
+        try:
+            self.bucket_name = self.connector.bucket_name
+        except AttributeError:
+            pass
```
[the remark this answers is not in the document]
Comment (ndokos) [reply server/lib/s3backup/__init__.py]: I was too lazy to change the check. Done now.
